### ISSUE TITLE
Non-localized assets seem broken

### ISSUE THREAD

**Issue by felixjung** [contentful/jekyll-contentful-data-import#46]
Hey,

I don't know what happened, but somehow over the past week our build broke. I can't recall updating any dependencies, but we keep getting
```
NoMethodError: undefined method `[]=' for nil:NilClass
```
when trying to fetch from our space.

Doing some digging with `pry` I was able to localize the source of this issue at [contentful.rb/lib/contentful/asset.rb#79](lib/contentful/asset.rb#L79):
```
@fields[default_locale][:file] = ::Contentful::File.new(file_json)
```
For some of our assets used in non-localized entries, `@fields[default_locale]` is `nil`, causing this line to fail. I don't understand the default locale is `nil`, as I would have thought the `FieldsResource` constructor should take care of initializing the default localization, no?

Is this maybe caused by us filtering for the locale we want to work on in the configuration of the Jekyll plugin?

```
  spaces:
    - de-DE:
        space: "xxx"
        access_token: "xxx"
        all_entries: true
        all_entries_page_size: 1000
        cda_query:
          include: 2
          limit: 100
          locale: "de-DE"
        client_options:
          api_url: "cdn.contentful.com"
        destination: "contentful"
```

I tried to quick-and-dirty patch this issue by either ensuring the `@fields[default_locale]` is initialized to an empty hash, if nil, or by directly setting `@fields[:file]` instead. Both ended up causing the following error:
```
NoMethodError: undefined method `url' for #<Hash:0x007fe13bb2c338>
  /usr/local/lib/ruby/gems/2.4.0/gems/jekyll-contentful-data-import-1.5.0/lib/jekyll-contentful-data-import/mappers/base.rb:119:in `map_asset'
```

Any idea what's going on?

**Comment by dlitvakb** [comment]
Heya @felixjung,

Can you check which versions of the CDA SDK and the Jekyll integration are you using? This may be related to the latest releases of both gems.

Cheers

**Comment by felixjung** [comment]
Some more hints. When debugging this, I noticed that the instances where `@fields[default_locale]` was `nil`, `@fields['de-DE']` was actually set to the exact value the line in question was going to set for `default_locale`. From the debugger repl:
``` ruby
From: /usr/local/lib/ruby/gems/2.4.0/gems/contentful-2.0.1/lib/contentful/asset.rb @ line 80 Contentful::Asset#create_files!:

    69: def create_files!
    70:   file_json = raw.fetch('fields', {}).fetch('file', nil)
    71:   return if file_json.nil?
    72:
    73:   is_localized = file_json.keys.none? { |f| %w(fileName contentType details url).include? f }
    74:   if is_localized
    75:     locales.each do |locale|
    76:       @fields[locale][:file] = ::Contentful::File.new(file_json[locale.to_s] || {})
    77:     end
    78:   else
    79:     require 'pry'; binding.pry
 => 80:     @fields[default_locale][:file] = ::Contentful::File.new(file_json)
    81:   end
    82: end

[5] pry(#<Contentful::Asset>)> default_locale
=> "en-US"
[6] pry(#<Contentful::Asset>)> @fields
=> {"de-DE"=>
  {:title=>"Irish flag",
   :file=>
    {"url"=>
      "//images.contentful.com/txhaodyqr481/4oqpo03lmwuY80Uy22Gy6K/63dc6a0b7ce3754f8b5b913baa99c8c7/ie.png",
     "details"=>{"size"=>451, "image"=>{"width"=>550, "height"=>275}},
     "fileName"=>"ie.png",
     "contentType"=>"image/png"}}}
[7] pry(#<Contentful::Asset>)> file_json
=> {"url"=>
  "//images.contentful.com/txhaodyqr481/4oqpo03lmwuY80Uy22Gy6K/63dc6a0b7ce3754f8b5b913baa99c8c7/ie.png",
 "details"=>{"size"=>451, "image"=>{"width"=>550, "height"=>275}},
 "fileName"=>"ie.png",
 "contentType"=>"image/png"}
```

**Comment by felixjung** [comment]
Hey @dlitvakb, so I might have run `bundle update`. I know these issues are happening with jekyll-contentful-data-import@1.5.1 and thus contentful.rb@2.0.1. I can't roll back because we rely on #28, which I think was not present in 1.4.3 (the last version before the update). We might have somehow got this working last week because we were running my fork.

**Comment by dlitvakb** [comment]
Looks like it's a simple configuration issue, just set the `client_options: { default_locale: 'de-DE' }` and you should be good to go.

Cheers

**Comment by felixjung** [comment]
Oooooh. Thanks. Wasn't aware of this setting. Is this new? Probably should've read the docs after updating ☺️

**Comment by dlitvakb** [comment]
Oh, no it isn't.

It's just that the new serializer depends on the client being actually properly configured and not just "guessing" things. Which makes it a "bit" (just one line) more complicated to use, but a lot more straight-forward to keep up with and maintain.

Cheers

**Comment by dlitvakb** [comment]
Maybe I should add it to the docs, in order to avoid confusion.

What do you think?

**Comment by felixjung** [comment]
Yeah, if this is required, I would definitely put it in the docs. I can't confirm whether what you suggest fixes my problem, but at least I've already spent ~2hrs debugging 😁

**Comment by dlitvakb** [comment]
Ok, please let me know as soon as you know if that fixed your issue.

If not, please open a support ticket and send us your Space ID and Access Token, and request for me, so I can debug it myself and see if there's a bug further than missing docs.

Cheers

**Comment by felixjung** [comment]
I've updated the configuration according to your suggestion and done some more research. Explicitly setting the `default_locale` in the config does not help. When I set the default locale to `en-US` (which is the actual default locale of our space) the problem remains the same and my output in the debugger looks like this:

``` yaml
[5] pry(#<Contentful::Asset>)> default_locale
=> "en-US"
[6] pry(#<Contentful::Asset>)> @fields
=> {"de-DE"=>
  {:title=>"Irish flag",
   :file=>
    {"url"=>
      "//images.contentful.com/txhaodyqr481/4oqpo03lmwuY80Uy22Gy6K/63dc6a0b7ce3754f8b5b913baa99c8c7/ie.png",
     "details"=>{"size"=>451, "image"=>{"width"=>550, "height"=>275}},
     "fileName"=>"ie.png",
     "contentType"=>"image/png"}}}
```

Note the value of `default_locale` (`en-US`) and the locale property set in the `@fields` hash (`de-DE`). So despite this asset not having set any localizations other than the default `en-US`, contentful.rb somehow localizes it to `de-DE`.

When I set `default_locale` to `de-DE`, as you suggest, the same problem occurs, but for a different asset and basically inverted:

```
[2] pry(#<Contentful::Asset>)> default_locale
=> "de-DE"
[3] pry(#<Contentful::Asset>)> @fields
=> {"en-US"=>
  {:title=>"German flag",
   :file=>
    {"url"=>
      "//images.contentful.com/txhaodyqr481/49qMVbLSDeCo8q68kquAO8/35353c0dcbfb414904622ae17930689b/de.png",
     "details"=>{"size"=>247, "image"=>{"width"=>550, "height"=>330}},
     "fileName"=>"de.png",
     "contentType"=>"image/png"}}}
```

contentful.rb now has already set the `en-US` value for the resolved field (as it usually should, because this is the space's default locale) but will now try to set `de-DE` because this is set as `default_locale` in the config.

The expected behavior for me here would be as follows:

1. Specify a filter for a given locale in the CDA query parameters (i.e., `de-DE`).
2. Contentful.rb will pull the German localizations of our content, falling back according to the fallback rules specified in Contentful.
3. Any entries, assets, etc. returned from contentful.rb will have the `en-US` field specified.

I get the feeling that contentful.rb gets confused because of the filter I set in the config. Is that possible?

I will note that we are having some issues with localized assets in the Contentful webapp, too. When picking assets to be inserted into a localized text field (say the German localization of a markdown text field), the file picker will not show previews for assets because somehow the localization fallbacks for assets don't work.

Given your previous comment, I will open a support ticket. Thanks for the help.
